### PR TITLE
Added FoldSubgraphEmptyInputs transformation

### PR DESCRIPTION
**Pull request by mbencer**
Ticket: 86099
- Added Folding of MultiSubgraph ops (Loop, TI, If) inputs to Constant if they are empty (at least one dimesion is zero)
- Relaxing `cloneWithNewDims` rules to handling zero dims
- Fixed comparing subgraph in test utils

### CODE REVIEW

File: src/plugins/intel_cpu/src/nodes/tensoriterator.cpp [hunk 1]
```diff
@@ -541,7 +541,7 @@ void TensorIterator::executeDynamicImpl(dnnl::stream strm) {
     sub_graph.ResetInferCount();
 
     bool continue_cond = initial_cond_check->getStatus();
-    int max_num_iter = trip_count_check->getStatus();
+     int max_num_iter = trip_count_check->getStatus();
```
Comment (luo-cheng2021) [inline src/plugins/intel_cpu/src/nodes/tensoriterator.cpp:544]: Should align correctly.

File: src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                         std::end(multi_subgraph_op_inputs),
+                         input,
+                         const_empty_replacement);
+            transformation_applied = true;
+        }
+
+        if (transformation_applied) {
+            multi_subgraph_op->set_arguments(multi_subgraph_op_inputs);
+            return true;
+        }
```
Comment (t-jankowski) [inline src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp:66]: Transformation happens only if `empty_inputs.size() > 0`, what should be clearly pointed by this code. Flag `transformation_applied` is not needed here.

File: src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+                // skip non-static shapes
+                const auto& in_shape = input.get_partial_shape();
+                if (in_shape.is_dynamic()) {
+                    return false;
+                }
+                return std::any_of(std::begin(in_shape), std::end(in_shape), [input](const ov::Dimension& dim) {
+                    if (dim.is_static() && dim.get_length() == 0) {
+                        return true;
+                    }
+                    return false;
```
Comment (t-jankowski) [inline src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp:47]:
```suggestion
                    return dim.is_static() && dim.get_length() == 0;
```

File: src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                });
+                return false;
```
Comment (t-jankowski) [inline src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp:49]: not needed.

**Review by t-jankowski — CHANGES_REQUESTED** [review]

File: src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp (hunk 2 above)
Comment (mbencer) [reply src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp]: good catch, thanks

File: src/plugins/intel_cpu/src/nodes/tensoriterator.cpp (hunk 1 above)
Comment (mbencer) [reply src/plugins/intel_cpu/src/nodes/tensoriterator.cpp]: done

File: src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp (hunk 4 above)
Comment (mbencer) [reply src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp]: done

**Review by luo-cheng2021 — APPROVED** [review]

File: src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                    return false;
+                }
+                // skip non-static shapes
+                const auto& in_shape = input.get_partial_shape();
+                if (in_shape.is_dynamic()) {
+                    return false;
+                }
+                return std::any_of(std::begin(in_shape), std::end(in_shape), [input](const ov::Dimension& dim) {
+                    return dim.is_static() && dim.get_length() == 0;
+                });
```
Comment (mitruska) [inline src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp:45]: Is `input` variable needed in the capture list?

File: src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                if (std::dynamic_pointer_cast<opset8::Constant>(input.get_node_shared_ptr())) {
+                    return false;
+                }
+                // skip non-static shapes
+                const auto& in_shape = input.get_partial_shape();
+                if (in_shape.is_dynamic()) {
+                    return false;
+                }
+                return std::any_of(std::begin(in_shape), std::end(in_shape), [input](const ov::Dimension& dim) {
+                    return dim.is_static() && dim.get_length() == 0;
```
Comment (mitruska) [inline src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp:44]: Immediate return after `in_shape.is_dynamic()` check gives quarantee, that all dimensions are static here, 
so check for `dim.is_static()` is redundant.

File: src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+                });
+            });
+
+        if (empty_inputs.size()) {
+            for (const auto& input : empty_inputs) {
+                const ov::Output<ov::Node> const_empty_replacement =
+                    std::make_shared<opset8::Constant>(input.get_element_type(), input.get_shape());
```
Comment (mitruska) [inline src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp:51]: Shouldn't we copy runtime info or friendly names for the new inputs?

File: src/tests/ie_test_utils/common_test_utils/graph_comparator.cpp [hunk 8]
```diff
@@ -313,7 +313,30 @@ class BackEdge {
...
+            return false;
+        }
+        const auto& param_shape = m_parameter->get_partial_shape();
+        const auto& result_shape = m_result->output(0).get_partial_shape();
+        const auto& param_rank = param_shape.rank();
+        const auto& result_rank = result_shape.rank();
+        if (!param_rank.compatible(result_rank)) {
+            return false;
+        }
+        if (param_rank.is_static() && result_rank.is_static()) {
```
Comment (mitruska) [inline src/tests/ie_test_utils/common_test_utils/graph_comparator.cpp:326]: Currently `result_and_parameter_match()` will return `true` also for pair of dynamic and static rank. Is it expected?

File: src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp (hunk 6 above)
Comment (mbencer) [reply src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp:44]: good catch, thanks

File: src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp (hunk 5 above)
Comment (mbencer) [reply src/common/transformations/src/transformations/common_optimizations/fold_subgraph_empty_inputs.cpp:45]: done, removed

File: src/tests/ie_test_utils/common_test_utils/graph_comparator.cpp (hunk 8 above)
Comment (mbencer) [reply src/tests/ie_test_utils/common_test_utils/graph_comparator.cpp:326]: I think so, especially param can have static rank, but we cannot calculate result rank. `param_rank.compatible(result_rank)` is adequate for me here

**Pull request merged**